doc/source/data/transforming-datasets.rst

```diff
@@ -301,6 +301,40 @@ The following output types are allowed for per-row UDFs (e.g.,
     :start-after: __writing_simple_out_row_udfs_begin__
     :end-before: __writing_simple_out_row_udfs_end__
 
+----------------------
+Configuring Batch Size
+----------------------
+
+:meth:`ds.map_batches() <ray.data.Dataset.map_batches>` is the canonical parallel
+transformation API for Datasets: it launches parallel tasks over the underlying Datasets
+blocks and maps UDFs over batches of data within those tasks, allowing the UDF to
+implement vectorized operations on batches. An important parameter to
+set is the ``batch_size``, which controls the size of batches provided to the UDF.
+
+.. literalinclude:: ./doc_code/transforming_datasets.py
+  :language: python
+  :start-after: __configuring_batch_size_begin__
+  :end-before: __configuring_batch_size_end__
+
+Increasing ``batch_size`` can result in faster execution by better leveraging SIMD
+hardware, reducing batch slicing overhead, and overall saturating CPUs/GPUs, but will
+also result in higher memory utilization, which can lead to out-of-memory failures.
+If encountering OOMs, decreasing your ``batch_size`` may help.
+
+.. note::
+  The default ``batch_size`` of ``4096`` may be too large for wide tables (many columns)
+  or datasets containing large images.
+
+Datasets will also bundle multiple blocks together for a single mapper task in order
+to better satisfy ``batch_size``, so if ``batch_size`` is a lot larger than your Dataset
+blocks (e.g. if your dataset was created with too large of a ``parallelism`` and/or the
+``batch_size`` is set to too large of a value), the number of parallel mapper tasks may
+be less than expected and you may be leaving some throughput on the table.
```

Comment:
Is this just referring to block coalescing? If so, won't the number of mapper tasks only be limited by `batch_size` and not the block size when parallelism is large?

```
num_tasks = num_rows/max(batch_size, block_size)
```

In other words, it's not really that the parallelism is too large, but still that the batch size is too large.

Reply:
>If so, won't the number of mapper tasks only be limited by batch_size and not the block size when parallelism is large?

Hmm the parallelism determines the block size, so that's a bit of a tautology.

>In other words, it's not really that the parallelism is too large, but still that the batch size is too large.

But the default parallelism can be set to a large value (relative to the batch size) automatically by Datasets, so if the user is relying the default parallelism but is manually setting the batch size (e.g. because they're tailored the batch size to what's needed to saturate a GPU for prediction), the `parallelism` may be the knob that they need to tune.

Reply:
@matthewdeng Does that make sense? Any ideas for how we can improve this section? Otherwise I'm planning on merging this.

Reply:
Ah reading this over I _think_ that I was referring to the last line:

> the number of parallel mapper tasks may be less than expected 

In which case lowering `parallelism` (or increasing `block_size`) won't increase the number of parallel tasks?

Reply:
This is more about making the transformation-time parallelism respect the read-time `parallelism` argument than increasing transformation-time parallelism, so maybe this doesn't need a call out?

The user expects the number of parallel mapper tasks to be what they give for read `parallelism`, but if `parallelism` is set too high, this can result in blocks that are much smaller than `batch_size`, and since we bundle blocks to try to achieve a full `batch_size` batch in our map tasks, the map parallelism can unexpectedly be a lot smaller than the read `parallelism`, which is what we're trying to call out here. All that decreasing `parallelism` would do in this case is result make the read-time parallelism more closely match the transformation-time parallelism.

Reply:
I see, my confusion was around what the user would consider "expected"  behavior, and (maybe due to my existing knowledge) I assumed they would be limited by batch size. I guess the other thing about this paragraph is that (unlike the following paragraph) there's no concrete action for the user to take.

I don't have a concrete suggestion to improve this, so feel free to consider this comment non-blocking.

Reply:
>I see, my confusion was around what the user would consider "expected" behavior, and (maybe due to my existing knowledge) I assumed they would be limited by batch size.

The commonly expected behavior I've heard from users is that their transformation parallelism is equal to their read parallelism, so this paragraph is mostly explaining that discrepancy for why those might not match.

>I guess the other thing about this paragraph is that (unlike the following paragraph) there's no concrete action for the user to take.

The action item of decreasing the batch size and/or decreasing the parallelism is somewhat implied by the "too large" comment, but since this is trying to alert about a possibly surprising discrepancy, not sure if we need a super explicit action item. 🤔 

>I don't have a concrete suggestion to improve this, so feel free to consider this comment non-blocking.

I'm going to leave this as-is for now so we have something in the docs to point to when talking about this parallelism discrepancy, but agreed that we might want to revisit this. I'm thinking that a lot of this feature guide will get reworked with the new execution planner + model, so we can revisit this at that time.